neo/io/nwbio.py

```diff
@@ -0,0 +1,810 @@
+            for signal in chain(segment.analogsignals, segment.irregularlysampledsignals):
+                if "nwb_electrode" in signal.annotations:
+                    elec_meta = signal.annotations["nwb_electrode"].copy()
+                    if elec_meta["name"] not in electrodes:
+                        # todo: check for consistency if the name is already there
+                        if elec_meta["device"]["name"] in devices:
+                            device = devices[elec_meta["device"]["name"]]
+                        else:
+                            device = nwbfile.create_device(**elec_meta["device"])
+                            devices[elec_meta["device"]["name"]] = device
+                        elec_meta.pop("device")
+                        electrodes[elec_meta["name"]] = nwbfile.create_icephys_electrode(
+                            device=device, **elec_meta
+                        )
+        return electrodes
+
+    def _write_segment(self, nwbfile, segment, electrodes):
+        # maybe use NWB trials to store Segment metadata?
+        for i, signal in enumerate(chain(segment.analogsignals, segment.irregularlysampledsignals)):
+            assert signal.segment is segment
+            if not signal.name:
+                signal.name = "%s : analogsignal%d" % (segment.name, i)
+            self._write_signal(nwbfile, signal, electrodes)
+
+        for i, train in enumerate(segment.spiketrains):
+            assert train.segment is segment
+            if not train.name:
+                train.name = "%s : spiketrain%d" % (segment.name, i)
+            self._write_spiketrain(nwbfile, train)
+
+        for i, event in enumerate(segment.events):
+            assert event.segment is segment
+            if not event.name:
+                event.name = "%s : event%d" % (segment.name, i)
+            self._write_event(nwbfile, event)
+
+        for i, epoch in enumerate(segment.epochs):
+            if not epoch.name:
+                epoch.name = "%s : epoch%d" % (segment.name, i)
+            self._write_epoch(nwbfile, epoch)
+
+    def _write_signal(self, nwbfile, signal, electrodes):
+        hierarchy = {'block': signal.segment.block.name, 'segment': signal.segment.name}
+        if "nwb_type" in signal.annotations:
+            timeseries_class = get_class(*signal.annotations["nwb_type"])
+        else:
+            timeseries_class = TimeSeries  # default
+        additional_metadata = {name[4:]: value
+                               for name, value in signal.annotations.items()
+                               if name.startswith("nwb:")}
+        if "nwb_electrode" in signal.annotations:
+            electrode_name = signal.annotations["nwb_electrode"]["name"]
+            additional_metadata["electrode"] = electrodes[electrode_name]
+        if timeseries_class != TimeSeries:
+            conversion, units = get_units_conversion(signal, timeseries_class)
+            additional_metadata["conversion"] = conversion
```

Comment:
Why only get conversion when the neurodata_type is not `TimeSeries`? [`TimeSeries`](https://nwb-schema.readthedocs.io/en/latest/format.html?highlight=TimeSeries#timeseries) and all its descendants contain a `conversion` field

Reply:
because in the general case of TimeSeries, there are no "expected" units, and so the conversion factor will always be 1.

Reply:
Hmm, well TimeSeries is required to have a unit, and although it is not specified, `conversion` still takes you from the values of the data to the physical values. This is particularly useful when the original data is stored as `int16` values.

Reply:
the scenario I have in mind here is where "generic" Neo data is being saved to NWB format without the annotations needed to determine a more suitable subclass of `TimeSeries`. In such a scenario, the Neo `AnalogSignal` definitely has a unit, and the conversion from the on-disk data to the physical values would have already taken place in the Neo IO class.

Reply:
added to #1013